lib/chef/provider/package/zypper.rb

```diff
@@ -139,7 +157,7 @@ def zypper_version
         end
 
         def zypper_package(command, global_options, *options, names, versions)
-          zipped_names = zip(names, versions)
+          zipped_names = source_files_exist? ? [new_resource.source] : zip(names, versions)
           if zypper_version < 1.0
             shell_out!("zypper", global_options, gpg_checks, command, *options, "-y", names)
           else
```

Comment:
there should just be one big conditional here which has `if new_resource.resource` followed by the zypper command to install all the sources (doesn't need to be flattened or anything as well since `shell_out!` will handle flattening arrays).  then the `else` side of the condition should be all the existing code.